evalml/tests/component_tests/detrender_tests/test_polynomial_decomposer.py

```diff
@@ -0,0 +1,400 @@
+
+    projected_seasonality = decomposer._build_seasonal_signal(
+        y_test,
+        single_period_seasonal_signal,
+        period,
+        frequency,
+    )
+
+    # Make sure that the function extracted the correct portion of the repeating, full seasonal signal
+    assert np.allclose(
+        full_seasonal_signal[projected_seasonality.index],
+        projected_seasonality,
+    )
+
+
+@pytest.mark.parametrize("X_num_time_columns", [0, 1, 2, 3])
+@pytest.mark.parametrize(
+    "X_has_time_index",
+    ["X_has_time_index", "X_doesnt_have_time_index"],
+)
+@pytest.mark.parametrize(
+    "y_has_time_index",
+    ["y_has_time_index", "y_doesnt_have_time_index"],
+)
+@pytest.mark.parametrize(
+    "time_index_specified",
+    [
+        "time_index_is_specified",
+        "time_index_not_specified",
+        "time_index_is_specified_but_wrong",
+    ],
+)
+def test_polynomial_decomposer_uses_time_index(
+    ts_data,
+    X_has_time_index,
+    X_num_time_columns,
+    y_has_time_index,
+    time_index_specified,
+):
+    X, y = ts_data
+
+    time_index_col_name = "date"
+    assert isinstance(X.index, pd.DatetimeIndex)
+    assert isinstance(y.index, pd.DatetimeIndex)
+
+    # Modify time series data to match testing conditions
+    if X_has_time_index == "X_doesnt_have_time_index":
+        X = X.reset_index(drop=True)
+    if y_has_time_index == "y_doesnt_have_time_index":
+        y = y.reset_index(drop=True)
+    if X_num_time_columns == 0:
+        X = X.drop(columns=[time_index_col_name])
+    elif X_num_time_columns > 1:
+        for addn_col in range(X_num_time_columns - 1):
+            X[time_index_col_name + str(addn_col + 1)] = X[time_index_col_name]
+    time_index = {
+        "time_index_is_specified": "date",
+        "time_index_not_specified": None,
+        "time_index_is_specified_but_wrong": "d4t3s",
+    }[time_index_specified]
+    decomposer = PolynomialDecomposer(time_index=time_index)
+
+    # The time series data has no time data
+    if (
+        X_num_time_columns == 0
+        and X_has_time_index == "X_doesnt_have_time_index"
+        and y_has_time_index == "y_doesnt_have_time_index"
+    ):
+        try:
+            output_X, output_y = decomposer.fit_transform(X, y)
+        except ValueError as e:
+            if (
+                "There are no Datetime features in the feature data and neither the feature or target data doesn't have Datetime index."
+                in str(e)
+            ):
+                pytest.xfail(
+                    "Must provide some sort of date time information to use this time series component.",
+                )
+
+    # The time series data has too much time data
+    if (
+        X_num_time_columns > 1
+        and time_index_specified == "time_index_not_specified"
+        and y_has_time_index == "y_doesnt_have_time_index"
+    ):
+        try:
+            output_X, output_y = decomposer.fit_transform(X, y)
+        except ValueError as e:
```

Comment:
Same question here, as well as below. All around, seems like we can refactor this section to something like:

```
if <first_case>:
   error_msg = "<first error message>"
elif <second_case>:
   error_msg = "<second error message>"
...
with pytest.raises(ValueError, match=error_msg):
   decomposer.fit_transform(X, y)
```

Reply:
big smart